### PR TITLE
feature/minimal-event-data-scraper

### PR DESCRIPTION
**Pull request by isaacna**
**Pull request recommendations:**
- [x] Name your pull request _your-development-type/short-description_. Ex: _feature/read-tiff-files_
- [ ] Link to any relevant issue in the PR description. Ex: _Resolves [gh-12], adds tiff file format support_
- [x] Provide context of changes.
 - PR for the minimal event data scraping methods. Doing this before making a prefect pipeline minimal event gather PR just to keep them from being too large
- Created a `MinimalEventData` object
- Added session support with a `MinimalSessionData` object, with a `MinimalEventAndSessionData` class that helps us handle the data for intermediate operations before matching sessions to events
- My session matching implementation only works under the assumption that there can be at most 2 sessions. I wrote it kinda fast and didn't realize there could potentially be more than 2, but I can just adjust that if we do find some edge cases
- [x] Provide relevant tests for your feature or bug fix.
 - Have done manual testing, parsing logic is essentially the same as other parsing methods
 - Paste showing the results of a MinimalEventGather (with sessions): https://pastebin.com/uBqpDPpn
- [ ] Provide or update documentation for any feature added by your pull request.

Thanks for contributing!

### CODE REVIEW

**Review by JacksonMaxfield — APPROVED** [review]
So, ignoring the failing builds due to deps here are my high level thoughts.

1. Please see my comment on "session" management
2. I am trying to figure out how to optimize all our collective working time. I think it would easier for @tohuynh to work on front-end, heck even more / other devs if we had a staging database with test data in it. So to resolve that, I think the best next steps would be for this work to be exported to it's own `staging-seattle` repo that contains the `Pulumi` stack init as a pip installed script from `cdp-deploy`, the basic web-app code from `cdp-instance`, and has `cdp-backend` as a dependency for pipeline stuff. I will have to talk to the `Pulumi` team in their Slack to see if it's possible to export a `Pulumi` stack init as a bin script or similar because then it can become: `pip install cdp-deploy` and `pip install cdp-backend` for all the backend deps and the typical `run_cdp_pipeline` for ingesting data into the staging database and then an `npm run watch` for launching and viewing the data in the web app?

(I would expect that we basically name change the current `seattle` repo to `seattle_v2` and `staging-seattle` to `seattle` when we are ready to release the new site)

So, with all that said, here are my (maybe our?) collective next steps?:
1. Create new `staging-seattle` repo
2. Work with `Pulumi` to see how possible it is to make a `pip` installable deployment script
3. Finish Database ORM in `cdp-backend`
4. Create base example data to fill staging database with (Likely just the `Model.Example()` from `cdp-backend` but maybe more)
5. We keep working on `cdp-deploy`, `cdp-backend`, and, `cdp-instance` with occasional PRs to `staging-seattle` when needed to pull in version updates from the others libraries.

I also wonder if we can overload GH Actions as the pipeline runner.... Just have GH Actions run some script every hour or so? I know there is a six hour timeout but I don't think we have any computation that lasts longer than maybe an hour? Which means that the secrets for each instance would just be GH Repo Secrets?

Pinging @nniiicc for thoughts too.

File: setup.py [hunk 1]
```diff
@@ -58,9 +59,17 @@
...
+    "google-api-core[grpc]==1.22.1",
     "google-cloud-storage>=1.28.1"
+    "google-api-python-client==1.10.0",
+    "google-auth==1.20.1",
+    "google-cloud-core==1.4.1",
+    "google-cloud-firestore==1.8.1",
+    "google-cloud-speech==1.3.2",
+    "google-cloud-storage==1.30.0",
+    "google-resumable-media==0.7.1",
 ]
```
[the remark this answers is not in the document]
Comment (isaacna) [reply setup.py]: I played myself, missed a comma and accidentally had some duplicate packages

File: cdptools/pipelines/minimal_event_data.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from datetime import datetime
+from typing import NamedTuple, Optional
+
+
+class MinimalEventData(NamedTuple):
+    body: str
+    event_datetime: datetime
+    video_uri: str
```
[the remark this answers is not in the document]
Comment (isaacna) [reply cdptools/pipelines/minimal_event_data.py:8]: Hm I see, I wasn't aware that there could be multiple sessions within the same event. How do we handle this currently?

[the remark this answers is not in the document]
Comment (isaacna) [reply cdptools/pipelines/minimal_event_data.py:8]: Also for point 2, you mean `staging-seattle` would primarily just hold city specific stuff (scrapers, etc.) and pull in all the generalized tools as needed?

File: setup.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply setup.py]: LOL. Well nice find.

File: cdptools/pipelines/minimal_event_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply cdptools/pipelines/minimal_event_data.py:8]: We currently don't handle them really at all except stating that they are the exact same event:

Morning session: https://councildataproject.github.io/seattle/#/events/c3d10348-5c4a-4228-8297-95174cd58233
Afternoon session: https://councildataproject.github.io/seattle/#/events/e80a6bb2-4036-44c5-9013-2d3663a07101

Check the details on both of them, there is no differences what-so-ever it's just transcript and video differences.

**Comment by JacksonMaxfield** [comment]
@isaacna 

> Also for point 2, you mean staging-seattle would primarily just hold city specific stuff (scrapers, etc.) and pull in all the generalized tools as needed?

Yes that would be exactly correct

File: cdptools/pipelines/minimal_event_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tohuynh) [reply cdptools/pipelines/minimal_event_data.py:8]: @JacksonMaxfield Given your suggestion about multiple sessions, how would filtering events by date work? 
We can't just put the date (without info on starting time) because it's possible to have sessions on different days?

[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply cdptools/pipelines/minimal_event_data.py:8]: Hmmmm I don't believe it is possible to have multiple sessions over multiple days. I think what I really meant was that sessions have a time and events have a date.

[the remark this answers is not in the document]
Comment (tohuynh) [reply cdptools/pipelines/minimal_event_data.py:8]: Then `MinimalEventData` needs a date field?

[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply cdptools/pipelines/minimal_event_data.py:8]: Yes. But really I would just have them all be datetimes. Where the EventDatetime is just the earliest session datetime. Which is why it probably isnt needed and be computed.

[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply cdptools/pipelines/minimal_event_data.py:8]: Basically, we can compute the event datetime from the provided session datetimes.

[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply cdptools/pipelines/minimal_event_data.py:8]: This is all operating under the assumption that we want to handle sessions. If we dont want to just yet then we dont need this.

[the remark this answers is not in the document]
Comment (isaacna) [reply cdptools/pipelines/minimal_event_data.py:8]: From a scraping perspective, how do we decide to group sessions to a single event? Do we just make an assumption that if a video url/webpage has the same body name and is within a similar time as others they are sessions of the same event? 

Or is there some data that actually links sessions together explicitly? I'm not that familiar with the seattle channel website, but for the 2 sessions you linked, I looked through the seattle channel html and didn't see anything that connects them

[the remark this answers is not in the document]
Comment (isaacna) [reply cdptools/pipelines/minimal_event_data.py:8]: Could there be a case where a body has 2 events in the same day, but they're about different topics so they aren't technically sessions?

[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply cdptools/pipelines/minimal_event_data.py:8]: I think the questions you just posed are enough for me to agree that we should, for now, only handle `Events` not `Sessions`.

**Comment by JacksonMaxfield** [comment]
@tohuynh if you are also happy with this I think you can merge @isaacna

I do think we will need to add an `Optional[List[MinutesItem]]` at some point down the line but that can be added after we get the basic pipeline and repo constructed that I described above.

File: cdptools/pipelines/minimal_event_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tohuynh) [reply cdptools/pipelines/minimal_event_data.py:8]: In those two linked events, they also share the same minutes items. In the future, that something to explore in trying to link sessions. I think we would have to make some sort of inference (probably can't guarantee 100% correct?) from body name, date, minutes items.

**Review by tohuynh — APPROVED** [review]

File: cdptools/pipelines/minimal_event_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tohuynh) [reply cdptools/pipelines/minimal_event_data.py:8]: on the seattlechannel, they also link the session in the description with: `Watch the afternoon session.`

[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply cdptools/pipelines/minimal_event_data.py:8]: We actually already do a text-diff compare to match seattle channel events to legistar events. If there are multiple events for the same body during the day we find the match by finding the least different minutes items list. Which is why I figure we could session stuff.

[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply cdptools/pipelines/minimal_event_data.py:8]: Because we are tossing this on the scraper side I think it would be okay to have sessions because as To pointed out, the seattle specific scraper could infer sessions by `regex`'ing "Watch the *** session" or similar.

[the remark this answers is not in the document]
Comment (isaacna) [reply cdptools/pipelines/minimal_event_data.py:8]: Sounds good, I'll work on adding logic to the parser for handling sessions and update it in this PR

**Comment by isaacna** [comment]
Some notes on these updates:
- The 2 session assumption could be limiting in the future, but since it's on the scraper side and I haven't seen edge cases (which I think would be held separately from our cdp-backend repo), I decided to roll with it. I can change it in the future if needed
- I decided to add "Minimal" for the new classes to not make the current cdptools code confusing, but will change it to just "event" and "session" when moving to cdp-backend
- The legistar attaching functioning only attaches body for now, but if we want to add `minutes_items` I can update that in a later pr (possibly after porting to cdp-backend)

**Review by JacksonMaxfield — CHANGES_REQUESTED** [review]
The actual implementation of `Event` and `Session` looks good to me! I am "Requesting changes" because I think you are no longer passing the minute details and such? We need those for votes, minutes items, matters, etc. No need to create objects for them yet. That is part of the `cdp-backend` migration but making sure our current pipeline doesn't break.

Note: When you transfer over this work, I would _only_ make models that directly correspond to how we expect data to be provided to us. `MinimalEventAndSessionData` for example, while useful, is only _immediately_ useful for the Seattle specific scrapper.

**Review by JacksonMaxfield — APPROVED** [review]
Oh, this is pointed at `release-3`. This won't break anything then. I thought this was pointed into `master`. I am fine with merging then haha. But we will need to address that `Optional[List[EventMinutesItem]]` when we port over to `cdp-backend

**Comment by isaacna** [comment]
Sounds good, I can create an issue for adding the minutes items. And minutes items are shared across sessions for the same event right?

**Comment by JacksonMaxfield** [comment]
> Sounds good, I'll note that on this [issue](https://github.com/CouncilDataProject/cdp-backend/issues/10). And minutes items are shared across sessions for the same event right?

For now, yes they are `EventMinutesItems`. I can imagine a future where we get fine grained enough for `SessionMinutesItems` though. But no need to worry about that just yet.

Feel free to merge whenever.
